### DISCUSSION TITLE
And what about Z-order indexing

### DISCUSSION THREAD

**Discussion by yudao** [sensedeep/dynamodb-onetable#257] [Ideas]
Hi,

First, a great, great, great big up for this awesome library. We have started using it and we found it very promising!

I just want to introduce a concept which could be very interesting, in particularly with one table architecture.

The [Z-order indexing](https://aws.amazon.com/blogs/database/z-order-indexing-for-multifaceted-queries-in-amazon-dynamodb-part-1/).

It will permit to create secondary indexes based on multi attributes for multifaceted queries. Maybe an attribute in the schema with will create automatically the sk when passing the table attributes needed for calculation?

It's just a open discussion to know if it will be interesting or not, because it's never magic :)

Regards,
Yu.

**Comment by mobsense** [comment]
Thank you for the kind words.

Z-order indexing is quite a mind bend. I'll have to re-read that blog post.

Can you suggest in detail how that would work?  What would the schema look like and the actual APIs?

**Reply by yudao** [reply]
In fact, if it's too mind bending, I agree ^^, the most important part is the summary:

> Z-order indexes allow developers to blend a set of attributes from a table’s items together to form a single, scalar, representative value. This value acts as the item’s sort key, allowing us to easily work within DynamoDB’s composite key uniqueness constraint. It also allows us to run queries against our data that benefit from multiple dimensional constraints, making Z-order indexes particularly suited to geospatial data and AWS IoT use cases.
> 
> Z-order indexes can outperform traditional indexes by one or more orders of magnitude in the best case, allowing you to save money on provisioned throughput. Depending on your query patterns, they might be able to replace entire sets of LSIs, reducing write overhead and cutting storage costs.

and the [second part](https://aws.amazon.com/blogs/database/z-order-indexing-for-multifaceted-queries-in-amazon-dynamodb-part-2/) which explains how to create a _z-index sk_.

Please keep in mind too:

> The core mechanism underlying Z-ordering is the ability to translate a tuple of several field values into a single number: that record’s Z-address. To achieve this, we simply interleave the bits of each field in the record.

My first thoughts about this type of indexing architecture, it's pretty simple on first approach.

1. Need an attribute on the entity like ```uuid: ulid|uuid``` but with an array of string type to pass all the attributes wanted into the z-index 
``` z-index: ['myAttr1','myAttr2','myAttr3','myAttr4']```
2. Automatically generates z-index by function on write event to create a z-index sk bytes based on selected attributes, the min/max value and according to the attributes type (integer, float, uff-8 text, timestamps, ...)

The 2 points cited above are not too difficult to create. The main problems I think is how to perform safeguard. I explain myself. If you pass any attribute for your z-index it can be less productive for indexation mechanism. It depends a lot of the type of attributes inside, the order and if can be boundaries values.

I started this discussion because, I know by experience that, we often need to perform queries on multiple attributes an we need to create GSI, that is not bad, but with a lot of duplicated data. Also, the Z-index indexation is very well suited for geolocation (which could be an entity primary type and in shadow use z-index sorting on just using longitude and latitude attributes in entity).

It will be a solid query search mechanism without creating too much GSI and duplicate data. And _de facto_ cost killing.

**Comment by mobsense** [comment]
Thank you. I'll dig into that.

There are some good articles on my blog about sparse indexes at https://www.sensedeep.com. Also, note the params {follow: true} where a GSI fetch can transparently fetch the full item. Costs an extra I/O but your GSI can the be KEYS_ONLY.

Lastly, remember to use item collections where if you organize your PK/SK, you can fetch a whole collection of related items. That is where DynamoDB really shines. I know it is often difficult to meet all uses cases with a single collection.

**Comment by yudao** [comment]
Thank you :)

I have already devoured your posts :). I understood sparse index.

It's not completely true, or I'm aware to listen your experience. Creating queries to find exactly what we are looking in less requests possible is quasi impossible, except using filtering (which querying lot more data than needed and returned). I eliminate always SCAN (excepts only for special admin actions in very few cases).

An exemple: 
```How  can I query an article between 2 creation date, in status "published" on a longitude X and latitude Y in a Multitenancy architecture?```
```How  can I query now the same article in status "published" on a longitude X in a Multitenancy architecture?```
```etc...```

You can't organise only with one PK and on composite SK because PK is the tenant id and the composite key can't be queried on any value of the composite key according to the order and due to lack of for example CONTAINS, ...

The only way we can is to generates GSI (with sparse index or not) and often duplicate data too or LSI.

But like I said I'm always aware of experiences especially yours! ;)

**Comment by mobsense** [comment]
If you use a composite sort key where it looks like

```
{ sk: { value: ${date}#${status}#${latx}#${laty} }
```

To query, use begins: `${Date.toISOString()}#published#${x}`

**Comment by yudao** [comment]
Yes if you search always with the same begin string. But in case you want to change order and query by Status first and latitude and longitude (facets) you can't.  You need to use GSI/LSI and often duplicate data. After talking regularly with AWS experts and by experience, create more than 5 GSI is not a good way. 

The z-index ordering, with it's limitation I agree, could let you query multi facets in one query and stop using too much GSI/LSI and data duplication.

I agree with you, if you create simple data you probably use all DynamoDB stuff without pain (in one table architecture). But, again by experience, when you have complicated use cases (e-commerce, Multitenancy, marketplace, geolocation, ...) it's the beginning of data duplication and too much GSI manipulations :)

**Reply by mobsense** [reply]
No disagreement here. I'm fuzzy on how to map attributes like strings into z-order components.

**Reply by yudao** [reply]
In the [part 2](https://aws.amazon.com/blogs/database/z-order-indexing-for-multifaceted-queries-in-amazon-dynamodb-part-2/) at section **Selecting binary representations for your attributes** you can see **UTF-8 text** which explain how to create z-order bytes from string.
